dashboard/app/models/course.rb

```diff
@@ -177,10 +177,26 @@ def assignable_info
     info
   end
 
-  # Get the set of valid courses for the dropdown in our sections table. This should
-  # be static data, but contains localized strings so we can only cache on a per
-  # locale basis
-  def self.valid_courses
+  # Get the assignable info for this course, substituting any alternate scripts
+  # based on any experiments this user has enabled, then update translations.
+  # @return AssignableInfo
+  def assignable_info_for_user(user)
+    info = ScriptConstants.assignable_info(self)
+    # ScriptConstants gives us untranslated versions of our course name, and the
+    # category it's in. Set translated strings here
+    info[:name] = localized_title
+    info[:category] = I18n.t('courses_category')
+    info[:script_ids] = scripts_for_user(user).map(&:id)
+    info
+  end
+
+  # Get the set of valid courses for the dropdown in our sections table. This
+  # should be static data for users without experiments enabled, but contains
+  # localized strings so we can only cache on a per locale basis.
+  def self.valid_courses(user = nil)
+    # Do not cache if the user might have an experiment enabled which puts them
+    # on an alternate script.
+    return Course.valid_courses_for_user(user) if CourseScript.has_any_experiment?(user)
```

Comment:
We could potentially cache by experiment if we wanted. Not sure that's necessary though.

Reply:
Not sure I feel strongly about this, but `User.has_course_experiment?` feels more natural to me than `CourseScript.has_any_experiment?(user)`. Thoughts?

Reply:
A user could have multiple experiments, so caching by experiment could add some complexity here. We could cache based on the set of experiments, or perhaps only cache if the user has a single experiment enabled.

Reply:
Ok. Let's leave this as is for now.

Reply:
Good point, although upon further reflection, experiment detection logic really belongs in the Experiment model. I've moved as much of it as I can there, extracting the experiment list to make it general.